src/providers/googleProvider.ts

```diff
@@ -1,59 +1,119 @@
 
-export interface RequestResult {
-  results: RawResult[];
-  status: string;
+// See https://developers.google.com/maps/documentation/javascript/geocoding
+// for insight into the GeocoderResult.
+interface RequestResult {
+  results: google.maps.GeocoderResult[];
 }
 
-export interface RawResult {
-  address_components: {
-    long_name: string;
-    short_name: string;
-    types: string[];
-  }[];
-  formatted_address: string;
-  geometry: {
-    location: LatLng;
-    location_type: string;
-    viewport: {
-      northeast: LatLng;
-      southwest: LatLng;
-    };
-  };
-  place_id: string;
-  plus_code: {
-    compound_code: string;
-    global_code: string;
-  };
-  types: string[];
+interface GeocodeError {
+  code: Exclude<google.maps.GeocoderStatus, google.maps.GeocoderStatus.OK>;
+  endpoint: 'GEOCODER_GEOCODE';
+  message: string;
+  name: 'MapsRequestError';
+  stack: string;
 }
 
+// TODO - given that we need to instantiate a GoogleProvider
+// with a geoCoder arg, what's the best way to actually deal with
+// this to make it as easy as possible for the end user?
+// We could perhaps instruct them to call loadGeocoder() and
+// pass the results as the second arg... or wrap it all in a promise
+// and return that instead?
 export default class GoogleProvider extends AbstractProvider<
   RequestResult,
-  RawResult
+  google.maps.GeocoderResult
 > {
-  searchUrl = 'https://maps.googleapis.com/maps/api/geocode/json';
+  geoCoder: google.maps.Geocoder;
 
-  endpoint({ query }: EndpointArgument) {
-    const params = typeof query === 'string' ? { address: query } : query;
-    return this.getUrl(this.searchUrl, params);
+  constructor(geoCoder: google.maps.Geocoder) {
+    super();
+    this._checkGeoCoder(geoCoder);
+    this.geoCoder = geoCoder;
   }
 
-  parse(result: ParseArgument<RequestResult>): SearchResult<RawResult>[] {
-    return result.data.results.map((r) => ({
-      x: r.geometry.location.lng,
-      y: r.geometry.location.lat,
-      label: r.formatted_address,
-      bounds: [
-        [r.geometry.viewport.southwest.lat, r.geometry.viewport.southwest.lng], // s, w
-        [r.geometry.viewport.northeast.lat, r.geometry.viewport.northeast.lng], // n, e
-      ],
-      raw: r,
-    }));
+  static async loadGeocoder(
+    options: ProviderOptions = {},
+  ): Promise<google.maps.Geocoder> {
+    return new Loader({
+      // The toString() is necessary to address a type mismatch -
+      // key can be string | number | boolean, but Google Maps
+      // expects it to be a string.
+      apiKey: options.params?.key?.toString() || '',
+      version: 'weekly',
+      // If a region is not specified, geocoding requests will bias toward the United States.
+      region: options.params?.region?.toString(),
+    })
+      .load()
+      .then((google) => {
+        return new google.maps.Geocoder();
+      });
+  }
+
+  endpoint({ query }: EndpointArgument): never {
+    throw new Error('Method not implemented.');
```

Comment:
@darrenklein You were on the right track here. If a method is declared as abstract in the parent, there's no way around it. I just think it's best practice to `throw` in these cases instead of simply returning a string in order to prevent any future errors from being "swallowed".

Reply:
Nope, this sounds good to me! I didn't know what the standard practice would be in this sort of situation, but that makes sense.